lib/generator/files.rb

```diff
@@ -0,0 +1,44 @@
+require 'pathname'
+
+module Generator
+  module Files
+    class ImmutableFile
+      attr_reader :filename, :repository_root
+      def initialize(filename:,repository_root: nil)
+        @filename = filename
+        @repository_root = repository_root
+      end
+
+      def to_s
+        File.read(filename) if File.exist?(filename)
+      end
+
+      def abbreviated_commit_hash
+        GitCommand.abbreviated_commit_hash(git_path, relative_filename)
+      end
+
+      private
+
+      def relative_filename
+        Pathname.new(filename).relative_path_from(Pathname.new(repository_root)).to_s
+      end
+
+      def git_path
+        File.join(repository_root, '.git')
+      end
+    end
+
+    class MutableFile < ImmutableFile
```

Comment:
This may be a case of misleading naming.

It's basically some files need to have a save method, and others don't. 
This way the intention of which files are meant to be `save`able is more clear.

It may not actually be important, and might just be a case of "just don't call save on the ones you don't want to save". 

Do you have any suggestions on how to handle this?

Doing the inheritance the other way around seems wrong if we need to define a save method which we then disable.

There's surely an elegant way of handling this hiding in there somewhere.